### PR TITLE
Updated to include Temperature Sensor

### PR DESCRIPTION
**Pull request by hpmax**
This update includes code to use the DS18B20 temperature sensor.  It has been tested and works.  It also includes code that is not currently used or tested that allows support for multiple devices.

### CODE REVIEW

File: configfile_parser.c [hunk 1]
```diff
@@ -1,123 +1,140 @@
...
-	}
-	else
-	{
-		// no config file used
-		return (0);
-	}
-}
+/*  
+	sensord - Sensor Interface for XCSoar Glide Computer - http://www.openvario.org/
+    Copyright (C) 2014  The openvario project
```
Comment (kedder) [inline configfile_parser.c:3]: Hm, why did the whole file have changed?

File: ds2482.c [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+{
+	// local variables
+	int fd;
+	
+	// try to open I2C Bus
+	fd = open("/dev/i2c-1", O_RDWR);
+	
+	if (fd < 0) {
+	   fprintf(stderr, "Error opening file: %s\n", strerror(errno));
+	   return 1;
```
Comment (kedder) [inline ds2482.c:55]: Please don't mix spaces and tabs. This project seems to use tabs as indentation characters. Let's keep it consistent.

**Review by kedder — CHANGES_REQUESTED** [review]
Correct me if I'm wrong, but to my understanding the temperature sensor is not part of "official" or "stefly" Openvario. How is it going to work when actual sensor is not installed? What of some other device installed on `/dev/i2c-1`? What if the device installed on different pin? 

IOW, how do we make this usable for all users?

Another issue with this PR: the diff treats several files as a complete rewrite. What happened to them? We should definitly not do a full reformatting of file as part of this PR. Also, let's keep indentation (and code style in general) consistent with the rest of the projet (tabs, no spaces).

Also, please don't submit PR with "fixup" commits (such as "Clean up"), squash them into commit that is being fixed.

**Comment by hpmax** [comment]
Andrey,

The entire temperature sensor support hardware is in both the "official" Openvario hardware and the "stefly" hardware.  Literally all I did (hardware wise) was to buy something like one of these:

[https://www.ebay.com/itm/193338692349](url)

and solder a 3.5mm stereo audio plug onto it to fit into the connector on the back of the unit.

Openvario does not communicate with the DS18B20 through the i2c bus,  It communicates with the DS2482 through the i2c bus, which in turn communicates with the DS18B20 through the "OneWire bus." The DS2482 is there, as far as I can tell, solely to communicate with the DS18B20 and is present in both the official and Stefly hardware.  If the DS18B20 is not there, or, for whatever reason, the DS2482 is missing, it should complain about it in one line on the log, and then just ignore both.  Now, if the DS2482 and another sensor is on the "One Wire" bus, that might cause an issue since I simply assume that if a device is present on the one wire bus, its a DS18B20.

As for the other concern... Yeah, sorry about that.  I was trying to ensure the indents lined up since it was previously kind of a mess.  Is there a tool to reformat the indents?

As for the large changes... I don't understand that either.  The only things that should have changed were Makefile (to add ds2482), d2482.c/ds2482.h, main.c, configfile_param.c, When codacy failed, I did start making changes to clean up the codacy failures (mainly moving the scope of variables).

**Comment by kedder** [comment]
> Openvario does not communicate with the DS18B20 through the i2c bus, It communicates with the DS2482 through the i2c bus, which in turn communicates with the DS18B20 through the "OneWire bus." 

Ah, that makes sense, I did not realize that. Thanks for explanation.

> Now, if the DS2482 and another sensor is on the "One Wire" bus, that might cause an issue since I simply assume that if a device is present on the one wire bus, its a DS18B20.

Alright, I've found [sensord description](https://www.openvario.org/doku.php?id=projects:series_00:software:sensord) on openvario.org, and yes, it looks like you are supposed to get DS18B20 temperature sensor. So we don't have to be super-flexible with supporting custom hardware here. That's good!

Now I wonder whether the temperature sensor has never been supported by sendord (despite docs in link above) or this PR is some kind of improved support for it?

> As for the other concern... Yeah, sorry about that. I was trying to ensure the indents lined up since it was previously kind of a mess. Is there a tool to reformat the indents?

I'm pretty sure most of code editors can do the reformatting. Just make sure you can control the result. Look at diffs before submitting, make sure you don't see changes that you did not intend to make.

> As for the large changes... I don't understand that either. The only things that should have changed were Makefile (to add ds2482), d2482.c/ds2482.h, main.c, configfile_param.c, When codacy failed, I did start making changes to clean up the codacy failures (mainly moving the scope of variables).

Could it be the accidental line ending change from UNIX (LF) to windows (CRLF) style? I'm pretty sure the onflicts in `main.*` files are because of this.

**Comment by hpmax** [comment]
I've been using emacs on my Ubuntu machine to edit.  Can you suggest an alternative, I'll try to reformat.  It's frustrating that the stuff got messed up like that.  I can't imagine I did anything to change it from LF to CRLF since the code never touched a Windows machine.

To the best of my knowledge, it was always intended that the DS18B20 would be supported, and the hardware provisions were all there, but there was no software to support it whatsoever.  As I mentioned before, the adapterboard includes a DS2482 whose sole purpose (as far as I can tell) is to communicate with the DS18B20 but prior to my code, there was no software to communicate with the DS2482.

While I can't vouch for my code being perfect, I believe I am simply fulfilling a planned future update from when the hardware was designed.  The DS18B20 has adjustable resolution (ranging from 1/2 of a degree C to 1/16th).  The more resolution you want the slower the conversion takes.  I am defaulting to 1/4 degree, but I added an optional line in the sensord.conf file to allow the user to control this (note that I should add bounds checking since its possible to put in a bogus number and it will accept it).

**Comment by kedder** [comment]
> I've been using emacs on my Ubuntu machine to edit. Can you suggest an alternative, I'll try to reformat. It's frustrating that the stuff got messed up like that. I can't imagine I did anything to change it from LF to CRLF since the code never touched a Windows machine.

Maybe it was the other way around (was formatted in windows-style). Which is not ideal, but if we want to reformat, let's do that in a separate PR, because it makes it impossible to review the changes.

I'm pretty sure emacs can convert spaces to tabs, but I'm no emacs expert, I used to use vim. In vim, you can do something like this:

```vim
set listchars=tab:▸\ ,trail:.,eol:¬  " to see spaces and tabs
set list " to see spaces and tabs
set ts=4 " or whatever value that makes tabs and spaces align
set noexpandtab " use tabs for indent
!retab " reformat the whole file
```

In vim, the `fileformat` option controls the line endings, you can save file with dos or unix by changing that option.


> To the best of my knowledge, it was always intended that the DS18B20 would be supported, and the hardware provisions were all there, but there was no software to support it whatsoever.

Got it!

> While I can't vouch for my code being perfect, I believe I am simply fulfilling a planned future update from when the hardware was designed. The DS18B20 has adjustable resolution (ranging from 1/2 of a degree C to 1/16th). The more resolution you want the slower the conversion takes. I am defaulting to 1/4 degree, but I added an optional line in the sensord.conf file to allow the user to control this (note that I should add bounds checking since its possible to put in a bogus number and it will accept it).

Makes sense!

**Comment by hpmax** [comment]
At this point, I'm inclined to kill my fork, re-fork the master distribution clone it and then revert the changes back into it.  Most of them are the addition of ds2482.c, I think there was a small section added to main, and one to the configfile parser, and of course Makefile.  It shouldn't take too long and then we are guaranteed to be starting clean.

Question is what Codacy issues should I fix that are outside my code update.  Currently, the errors seem to be:

1) Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).
This is in my code, and frankly, I don't understand it.  The code is: 
if (read(sensor->fd,&data,1)==0) { //Read the status register

&data is a pointer to a character, so it should never read past the end of the buffer because it's only reading 1 byte.

2) Doesn't like usleep... Apparently nanosleep is more accurate, but I doubt any of this is time critical.

3) Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).
if ((send(sock, s, strlen(s)+1, 0)) < 0) fprintf(stderr, "send failed\n");

In this case, if s isn't properly terminated with a zero at the end, it could do bad things because strlen would report the wrong number and it could write out data not in s.  However, s is generated by sprintf (I think), and the length is always significantly shorter than the length of s, so this should never be a problem.  I think there is an open issue about whether or not the terminating 0 should be transmitted.  This apparently transmits it thanks to the strlen(s)+1 instead of strlen(s).  It probably should not be transmitted

4) Ensure that umask is given most restrictive possible setting (e.g., 066 or 077) (CWE-732).
umask(0);

Isn't that the most restrictive possible setting?  Weird message.

**Comment by kedder** [comment]
> At this point, I'm inclined to kill my fork, re-fork the master distribution clone it and then revert the changes back into it. Most of them are the addition of ds2482.c, I think there was a small section added to main, and one to the configfile parser, and of course Makefile. It shouldn't take too long and then we are guaranteed to be starting clean.

Strictly speaking you don't have to kill the fork. You can force-push to the same branch in your fork and this PR will get updated. Closing PR and opening new one is generally not encouraged, because all the reviews and discussons are effectively lost that way. But in this case I guess it doesn't matter that much.

> Question is what Codacy issues should I fix that are outside my code update. 

I think we should take Codacy with the grain of salt. It tries to report all potential issues regardless whether they make sense or not in a given context. We should just review them, fix those that make sense and leave the irrelevant ones as is. I think there should be some sort of UI to mark issues as "irrelevant" in codacy, but it is probably only available to @linuxianer99.

**Pull request closed without merging**